llvm/test/CodeGen/AIE/aie2/schedule/loopaware/short-hwloop-resources.mir

```diff
@@ -0,0 +1,191 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py
+
+# This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+#
+# (c) Copyright 2024 Advanced Micro Devices, Inc. or its affiliates
+
+# RUN: llc --mtriple=aie2 --run-pass=postmisched --aie-interblock-scoreboard=0 \
+# RUN:  %s -o - | FileCheck %s --check-prefix=NO-LOOPAWARE-SCOREBOARD
+# RUN: llc --mtriple=aie2 --run-pass=postmisched --aie-interblock-scoreboard=1 \
+# RUN:  %s -o - | FileCheck %s --check-prefix=LOOPAWARE-SCOREBOARD
+
+# Currently, the loop-aware scheduler only allows resources/FuncUnits to stick
+# out of the loop into the next iteration. It does not allow them to stick out
+# for more than one iteration.
+
+# This test here shows potential issues with epilogue scheduling if we were to
+# change the requirements stated above. In the NO-LOOPAWARE-SCOREBOARD checks,
+# we show that resource conflicts between the loop and the epilogue are not
+# correctly avoided for tiny loops. I.e. loops with less than
+# getConflictHorizon() cycles.
+
+
+# In the epilogue, CM_WM_PORT is booked by VLDA in cycle 6.
+# As the loop will potentially execute multiple times, its VLDA_UPS instruction
+# can then book CM_WM_PORT in cycle 7,6,5,4,3,2,1,0 of the epilogue.
+# There should then be two NOPs between the loop and the epilogue.
+---
+name:            epilogue_c6_blocked
+alignment:       16
+tracksRegLiveness: true
+body:             |
+  ; NO-LOOPAWARE-SCOREBOARD-LABEL: name: epilogue_c6_blocked
+  ; NO-LOOPAWARE-SCOREBOARD: bb.0:
+  ; NO-LOOPAWARE-SCOREBOARD-NEXT:   successors: %bb.1(0x80000000)
+  ; NO-LOOPAWARE-SCOREBOARD-NEXT: {{  $}}
+  ; NO-LOOPAWARE-SCOREBOARD-NEXT: bb.1:
+  ; NO-LOOPAWARE-SCOREBOARD-NEXT:   successors: %bb.1(0x40000000), %bb.2(0x40000000)
+  ; NO-LOOPAWARE-SCOREBOARD-NEXT:   liveins: $s1, $p2
+  ; NO-LOOPAWARE-SCOREBOARD-NEXT: {{  $}}
+  ; NO-LOOPAWARE-SCOREBOARD-NEXT:   $cm2, $p2 = VLDA_UPS_S32_S8_ag_pstm_nrm_imm $s1, killed $p2, 32, implicit-def $srups_of, implicit $crsat
+  ; NO-LOOPAWARE-SCOREBOARD-NEXT:   PseudoLoopEnd %bb.2, %bb.1
+  ; NO-LOOPAWARE-SCOREBOARD-NEXT: {{  $}}
+  ; NO-LOOPAWARE-SCOREBOARD-NEXT: bb.2:
+  ; NO-LOOPAWARE-SCOREBOARD-NEXT:   liveins: $p2, $s2
+  ; NO-LOOPAWARE-SCOREBOARD-NEXT: {{  $}}
+  ; NO-LOOPAWARE-SCOREBOARD-NEXT:   $amlh5, $p2 = VLDA_dmw_lda_am_ag_pstm_nrm_imm killed $p2, 2016
```

Comment:
Just for clarification, we booked some cycles to avoid loop-caried resource conflicts to the epilogue, right? In this case the comment mention two nops here.  I think we have more than enough nops here....

Reply:
Not exactly, there should be two NOPs before `VLDA_dmw_lda_am_ag_pstm_nrm_imm` to make sure the epilogue cannot have resource conflicts with **any** of the previous iterations of the loop. They are not there because in epilogue scheduling we only account for the **last** iteration of the loop. That is safe to do in the normal case because the loop scheduling itself ensures that scoreboard resources never stick out of more than one iteration. it's just that the scoreboard is disabled for the `NO-LOOPAWARE-SCOREBOARD` label, just to show how epilogue scheduling can get complicated :)

Reply:
Understood! thank you very much!